src/python/pants/java/nailgun_io.py

```diff
@@ -205,6 +209,73 @@ def __init__(self, in_fds, sock, chunk_types, chunk_eof_type, buf_size=None, sel
   def _assert_aligned(self, *iterables):
     assert len({len(i) for i in iterables}) == 1, 'inputs are not aligned'
 
+  def _handle_closed_input_stream(self, fileno):
+    # We've reached EOF.
+    try:
+      if self._chunk_eof_type is not None:
+        NailgunProtocol.write_chunk(self._socket, self._chunk_eof_type)
+    finally:
+      try:
+        os.close(fileno)
+      finally:
+        self.stop_reading_from_fd(fileno)
+
+  def stop_reading_from_fd(self, fileno):
+    self._in_fds.remove(fileno)
+
+  def run(self):
+    while self._in_fds and not self.is_stopped:
+      readable, _, errored = select.select(self._in_fds, [], self._in_fds, self._select_timeout)
+      self.do_run(readable, errored)
+
+  def do_run(self, readable, errored):
```

Comment:
Nit: Call `readable` instead `readables` to reflect its plurality?

Reply:
Done